tests/cpp/test_multidevice_transformer.cpp

```diff
@@ -224,14 +237,73 @@ std::vector<TensorView*> mlp(
   return {linear1, gelu, linear2, dropout};
 }
 
+std::vector<TensorView*> mha(
+    TensorView* x,
+    TensorView* w0,
+    TensorView* b0,
+    TensorView* w1,
+    TensorView* b1,
+    DeviceMesh& mesh,
+    DataType dtype) {
+  // Linear 1
+  TensorView* mm = matmul(x, w0);
+  TensorView* proj_bias_bcast = broadcast(b0, {false, true, false});
+  TensorView* qkv1 = add(mm, proj_bias_bcast);
+  // Forming the q,k,v vectors:
+  auto D = w0->axis(0)->extent()->value().as<int64_t>();
+  TensorView* qkv = reshape(qkv1, {D, B * S, 3 * E / D}, {D, B, S, 3 * E / D});
+  std::vector<TensorView*> qkv_reshaped = {};
+  for (auto i : c10::irange(3)) {
+    TensorView* tv_slice =
+        slice(qkv, {0, 0, 0, E / D * i}, {D, B, S, E / D * (i + 1)});
+    TensorView* tv_reshape =
+        reshape(tv_slice, {D, B, S, E / D}, {D, B, S, H / D, E / H});
+    TensorView* tv_trans = transpose(tv_reshape, 2, 3);
+    TensorView* tv_cast = castOp(dtype, tv_trans);
+    qkv_reshaped.push_back(tv_cast);
+    // Explicitly shard qkv before calling SDPA node
+    for (auto tv : {tv_slice, tv_reshape, tv_trans, tv_cast}) {
+      tv->setDeviceMesh(mesh);
+      tv->axis(0)->parallelize(ParallelType::DIDx);
+    }
+  }
+  // SDPA
+  SdpfaFwdResult sdpa = sdpfa_fwd(
+      qkv_reshaped[0],
+      qkv_reshaped[1],
+      qkv_reshaped[2],
+      IrBuilder::create<Val>(kSdpaProb),
+      IrBuilder::create<Val>(true),
+      IrBuilder::create<Val>(kSdpaScale));
+  TensorView* sdpa_output = sdpa.output;
+  // Linear projection
+  TensorView* sdpa_transpose = transpose(sdpa_output, 2, 3);
+  TensorView* sdpa_reshape =
+      reshape(sdpa_transpose, {D, B, S, H / D, E / H}, {D, B * S, E / D});
+  TensorView* mm2 = matmul(sdpa_reshape, w1);
+  TensorView* mm2_ar = sum(mm2, {0}); // allreduce
+  TensorView* b1_bcast = broadcast(b1, {true, false});
+  TensorView* linear2 = add(mm2_ar, b1_bcast);
+  // Dropout
+  TensorView* dropout = replicated_dropout(linear2, kDropoutProb, mesh);
+
+  for (auto tv : {x, b1, mm2_ar, linear2, dropout}) {
+    tv->setDeviceMesh(mesh);
+  }
+  for (auto tv : {w0, b0, w1, proj_bias_bcast, mm, mm2, qkv, sdpa_output}) {
+    tv->setDeviceMesh(mesh);
+    tv->axis(0)->parallelize(ParallelType::DIDx);
+  }
+  return {qkv, sdpa_output, linear2, dropout};
+}
+
 std::vector<TensorView*> mlp_backwards(
     TensorView* grad,
     TensorView* x,
     TensorView* mask,
     TensorView* w0,
     TensorView* b0,
     TensorView* w1,
-    Fusion* fusion,
     DeviceMesh& mesh,
```

Comment:
```suggestion
    const DeviceMesh& mesh,
```
and several other places in this file.